[Tablet Orders] Stop toggling in the product selector

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12135 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Previously, tapping an item in the Product Selector _toggled_ the selected state in its own view model and told its parent view model that the selection state had changed, not what it had changed to.

This meant that if the selection state in the ProductSelectorViewModel and the code which provided the `onProductSelectionStateChanged` closure had divergent states for what products were selected, the UI would behave unpredictably, with a checked item being removed from an order, or unchecking an item adding it to the order.

This PR changes the code to specify the selection state, and avoid toggle actions.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Perform the following tests with `sideBySideViewForOrderForm` feature flag set to true, then repeat with it set to false.

### Order Creation

1. Launch the app on a large iPhone (in portrait) or iPad (in a small split or slideover.)
2. Navigate to the `Orders` tab and tap `+`; observe that the Order Form opens without a product selector.
3. Tap `+ Add Products`.
4. Select some products, tap `Add Products`; observe that they're added to the order
5. Tap `+` at the top of the Products section.
6. Deselect one or more products.
7. Close the product selector without confirming; observe that the order doesn't change.
8. Tap `+` again, observe that the selections match the order.
9. Once again deselect the products you previously deselected.
10. Confirm the change; observe that the order is updated correctly.


### Coupons
1. Navigate to `Menu > Coupons`.
2. Add a new coupon.
3. Tap `All Products` to open the product selector.
4. Add some products and confirm the selection.
5. Re-open the product selector, deselect, and close without saving
6. Re-open the product selector; observe that the state is correct
7. Once again deselect the products
8. Confirm the change; observe that the coupon is updated correctly.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
12135 Specify selection state, don’t toggle

Toggling the selection state can lead to inconsistencies between the Order details and the product selector. For example, when out of sync changing the visual state to show a checkmark can actually remove the item from the order.

This replaces toggling with specifying the desired selection state, matching the visual style.

joshheald committed:
12135 Use presentationStyle for Add Products visibility

Prior to this change, the Coupons product selector didn’t have an Add Products button when the feature flag was true.

Making the environment var optional allows us to be more accurate in consulting it; if the view isn’t embedded in an AdaptiveModalContainer, it will be `nil`.

Previous issues with `onChange` may still apply, but we don’t look for that notification here.

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
12135 Notify with selection state, not only change

Previously, we only notified that the selection state had changed, not what it changed to.

This commit updates that to pass the new selection state to the parent view model, for easier consistency between the selector and its caller.

joshheald committed:
Merge branch 'trunk' into issue/12135-stop-toggling-in-the-product-selector

joshheald committed:
Revert unintended commit of enabled feature flag

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -132,11 +132,11 @@ final class ProductSelectorViewModel: ObservableObject {
     /// Closure to be invoked when a product is selected or deselected
     ///
-    private let onProductSelectionStateChanged: ((Product) -> Void)?
+    private let onProductSelectionStateChanged: ((Product, Bool) -> Void)?
```
General question: What do you think about using a `Bool` to represent "selection state" on these signatures, when the `SelectedState` enum has 3 states and not 2? (.selected, .notSelected, .partiallySelected). Would it be worth to change these to something like the following?
```
private let onProductSelectionStateChanged: ((Product, ProductRow.SelectedState) -> Void)?
```
That said, I'm more inclined to - in favour of simplification- of removing the `.partiallySelected` case, which only seems to modify the style for the parent row when a variation is selected, making a row being either selected or unselected.

joshheald replied:
Yeah, I considered that, and it sounds like we came to the same confusion. 

Since we never actually change a product's selection state to `partiallySelected` except as a UI provision. I don't think there's a need to inform the parent view model of that UI-only state, and it makes the code more complex if we do.

If we ever need to track that as an analytic or do something else in the parent view, we can change to use the enum in future.

iamgabrielma reviewed (approved):
That's a very welcome change, thanks for updating this behaviour.

LGTM! :shipit:
